utils/log-collector

```diff
@@ -0,0 +1,101 @@
+
+now=$(date +"%F")
+outputFormat="${outputFormat:=text}"
+
+# Create a temporary directory to store our logs
+pwd=$(pwd)
+dir=$(mktemp -d "$pwd"/buildkite-logs-"$now")
+
+echo "Collecting logs for instance $instanceId"
+
+# Collect the logs for the instance
+aws logs get-log-events --log-group-name /buildkite/buildkite-agent --log-stream-name "$instanceId" --output "$outputFormat" >> "$dir"/buildkite-agent-"$instanceId".log
+aws logs get-log-events --log-group-name /buildkite/system --log-stream-name "$instanceId" --output "$outputFormat" >> "$dir"/system-"$instanceId".log
+aws logs get-log-events --log-group-name /buildkite/lifecycled --log-stream-name "$instanceId" --output "$outputFormat" >> "$dir"/lifecycled-"$instanceId".log
+aws logs get-log-events --log-group-name /buildkite/docker-daemon --log-stream-name "$instanceId" --output "$outputFormat" >> "$dir"/docker-daemon-"$instanceId".log
+
+# CloudWatch stores the creationTime as milliseconds since epoch UTC. 
+# Since macOS doesn't support calculating the time since epoch in milliseconds, we don't have perfect accuracy on this, but it should be as close to 24 hours as possible
+# Currently only collects the last 24h of logs, but it might be good to be able to specify a different value for this.
+logAge=$(date -v-1d "+%s000")
+
+# Collect some information about the Lambda function and associated log groups
+# Change the timeframe to be dynamically set with a default of 24h. Currently collects the last 24 hours of Lambda logs
+lambdaFunction=$(aws cloudformation describe-stack-resources --stack-name "$stackName" --logical-resource-id Autoscaling --query "StackResources[*].PhysicalResourceId" --output text | sed -nr 's/.*stack\/(.*)\/.*/\1/p')
+lambdaLogGroup=$(aws cloudformation describe-stack-resources --stack-name "$lambdaFunction" --logical-resource-id LogGroup --query "StackResources[*].PhysicalResourceId" --output text)
+logStreams=$(aws logs describe-log-streams --log-group-name "$lambdaLogGroup" --order-by LastEventTime --query "logStreams[?creationTime > \`$logAge\`] | [*].logStreamName" --output text | tr -s '[:space:]' '\n')
+
+# Iterate through all of the log streams we collected from the lambda log group
+
+echo "Collecting Lambda logs for $lambdaFunction"
+for i in $logStreams
+do
+    fileName=$(echo "$i" | sed -r 's/\//-/g')
+    aws logs get-log-events --log-group-name "$lambdaLogGroup" --log-stream-name "$i" --output "$outputFormat" >> "$dir"/lambda-"$fileName".log
+done
+
+# Collect ASG activity for the scaling group
+autoScalingGroup=$(aws cloudformation describe-stack-resources --stack-name "$stackName" --logical-resource-id AgentAutoScaleGroup --query "StackResources[*].PhysicalResourceId" --output text)
+echo "Collecting ASG activity for $autoScalingGroup"
+aws autoscaling describe-scaling-activities --auto-scaling-group-name "$autoScalingGroup" >> "$dir"/asg-activities-"$stackName".log
+
+# Make an archive with the logs
+zip -jq buildkite-logs-"$now".zip "$dir"/*.log
```

Comment:
I wonder if you'll have trouble with `zip` availability.

macOS has it, but e.g. The minimal linux docker containers don't tend to have it:

```
pda@paulbookpro ~/bk/buildkite $ for img in debian ubuntu archlinux centos alpine; do echo -n "$img: "; docker run --rm $img sh -c "zip"; done
debian: sh: 1: zip: not found
ubuntu: sh: 1: zip: not found
archlinux: sh: line 1: zip: command not found
centos: sh: zip: command not found
alpine: sh: zip: not found
```

Tar might be a safer assumption…


```
pda@paulbookpro ~/bk/buildkite $ for img in debian ubuntu archlinux centos alpine; do echo -n "$img: "; docker run --rm $img sh -c "tar --version | head -n1"; done
debian: tar (GNU tar) 1.34
ubuntu: tar (GNU tar) 1.30
archlinux: tar (GNU tar) 1.34
centos: tar (GNU tar) 1.30
alpine: tar (busybox) 1.34.1
```

But maybe people are mostly doing this from fully-optioned macOS / Linux machines that are likely to have `zip`.